### PR TITLE
posix: Implement popen()

### PR DESCRIPTION
**Pull request by mintsuki**

### CODE REVIEW

File: options/posix/generic/posix_stdio.cpp [hunk 1]
```diff
@@ -18,9 +24,103 @@ int pclose(FILE *) {
...
+
+	if (!mlibc::sys_fork || !mlibc::sys_close || !mlibc::sys_dup2 || !mlibc::sys_execve
+			|| !mlibc::sys_sigprocmask || !mlibc::sys_sigaction || !mlibc::sys_pipe) {
+		MLIBC_MISSING_SYSDEP();
+		errno = ENOSYS;
+		return nullptr;
+	}
+
+#if __MLIBC_POSIX_OPTION
+	pthread_testcancel();
```
Comment (Geertiebear) [inline options/posix/generic/posix_stdio.cpp:40]: I don't see where ``popen()`` is specified as a cancellation point in POSIX.

**Review by Geertiebear — CHANGES_REQUESTED** [review]
Is it possible to also make a small test case for ``popen()``? Something like executing ``echo "test"`` or something like that and checking if the correct output is there?

File: options/posix/generic/posix_stdio.cpp (hunk 1 above)
Comment (Dennisbonke) [reply options/posix/generic/posix_stdio.cpp:40]: [This[(https://pubs.opengroup.org/onlinepubs/000095399/functions/xsh_chap02_09.html#tag_02_09_05_02) says that it _may_ be a cancellation point so I think that is correct.

Comment (Geertiebear) [reply options/posix/generic/posix_stdio.cpp:40]: Ah I completely read over that, my bad. I'm still not sure we should include ``pthread_testcancel()`` however, since POSIX says it _can_ be a cancellation point to cover implementations that use blocking syscalls to implement it. In general library functions don't have to explicitely test for cancellations, since the sysdeps should already check for it appropriately.

E.g. nowhere in the ``fputc()`` implementation do we check using ``pthread_testcancel()``, but since it calls into ``read()``, it automatically becomes a cancellation point.

File: options/posix/generic/posix_stdio.cpp [hunk 2]
```diff
@@ -2,11 +2,17 @@
...
 #include <stdlib.h>
+#include <unistd.h>
 
 #include <bits/ensure.h>
 #include <mlibc/debug.hpp>
 #include <mlibc/file-io.hpp>
 #include <mlibc/posix-file-io.hpp>
+#include <mlibc/posix-sysdeps.hpp>
+
+#if __MLIBC_POSIX_OPTION
```
[the remark this answers is not in the document]
Comment (mintsuki) [reply options/posix/generic/posix_stdio.cpp:13]: The `#if`? Because the include did not seem unnecessary.

File: options/posix/generic/posix_stdio.cpp [hunk 3]
```diff
@@ -18,9 +24,103 @@ int pclose(FILE *) {
...
 
-FILE *popen(const char*, const char *) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+FILE *popen(const char *command, const char *typestr) {
+	bool is_write;
+	pid_t child;
+	FILE *ret = nullptr;
+
+	if (!mlibc::sys_fork || !mlibc::sys_close || !mlibc::sys_dup2 || !mlibc::sys_execve
```
[the remark this answers is not in the document]
Comment (mintsuki) [reply options/posix/generic/posix_stdio.cpp:32]: I just did what `system()` already did.

File: options/posix/generic/posix_stdio.cpp [hunk 4]
```diff
@@ -18,9 +24,103 @@ int pclose(FILE *) {
...
+		errno = ENOSYS;
+		return nullptr;
+	}
+
+#if __MLIBC_POSIX_OPTION
+	pthread_testcancel();
+#endif // __MLIBC_POSIX_OPTION
+
+	if (!command) {
+		return nullptr;
```
[the remark this answers is not in the document]
Comment (mintsuki) [reply options/posix/generic/posix_stdio.cpp:44]: I did what `system()` did.

File: options/posix/generic/posix_stdio.cpp (hunk 1 above)
Comment (mintsuki) [reply options/posix/generic/posix_stdio.cpp:40]: Again, I did what `system()` did :^)

File: options/posix/generic/posix_stdio.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/posix/generic/posix_stdio.cpp:13]: Yeah but you don't need to do the ``#if`` if we are already in the POSIX option.

File: options/posix/generic/posix_stdio.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/posix/generic/posix_stdio.cpp:32]: Guess that's fine then.

File: options/posix/generic/posix_stdio.cpp (hunk 1 above)
Comment (Geertiebear) [reply options/posix/generic/posix_stdio.cpp:40]: ``system()`` has the requirement that it _has_ to be a cancellation point, ``popen()`` doens't have this requirement. So imo it's not needed and can be removed.

File: options/posix/generic/posix_stdio.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Geertiebear) [reply options/posix/generic/posix_stdio.cpp:44]: In the man pages for ``system()`` it explicitely mentions the case ``command == NULL`` but the man page for ``popen()`` doesn't, so I don't think any special handling needs to be done for this case (although maybe a test case to verify it doesn't cause a null-deref would be good).

**Comment by mintsuki** [comment]
Done.

File: options/posix/generic/posix_stdio.cpp [hunk 5]
```diff
@@ -18,9 +20,98 @@ int pclose(FILE *) {
...
+	FILE *ret = nullptr;
+
+	if (!mlibc::sys_fork || !mlibc::sys_close || !mlibc::sys_dup2 || !mlibc::sys_execve
+			|| !mlibc::sys_sigprocmask || !mlibc::sys_sigaction || !mlibc::sys_pipe) {
+		MLIBC_MISSING_SYSDEP();
+		errno = ENOSYS;
+		return nullptr;
+	}
+
+	__ensure(command);
```
Comment (Geertiebear) [inline options/posix/generic/posix_stdio.cpp:35]: But now we're still checking for ``command``

File: options/posix/generic/posix_stdio.cpp [hunk 6]
```diff
@@ -18,9 +20,98 @@ int pclose(FILE *) {
...
+    __ensure(typestr);
```
Comment (Geertiebear) [inline options/posix/generic/posix_stdio.cpp:36]: A ``nullptr`` ``typestr`` should cause ``errno = EINVAL``.

**Review by Geertiebear — COMMENTED** [review]
Two more comments, for the rest LGTM.

**Review by Geertiebear — APPROVED** [review]
LGTM now!
